### ISSUE TITLE
Date offset in observations compared to simulation results

### ISSUE THREAD

**Issue by wouterjdb** [equinor/flownet#207]
Using  `time_ag_open_location`

![image](https://user-images.githubusercontent.com/9119793/95589216-bb381200-0a44-11eb-99dd-1a44523a102d.png)

**Comment by olelod** [comment]
Looking into some simulation files.

Clip from HISTORIC_SCHEDULE.inc:
![HIST_SCHED](https://user-images.githubusercontent.com/46521401/96582652-e1c73a00-12db-11eb-9cbc-f588618744c0.JPG)

Clips from observations.ertobs:
![ertobs](https://user-images.githubusercontent.com/46521401/96582732-03282600-12dc-11eb-9d60-0e2fc85e9759.JPG)

WOPR:D-1H and WOPRH:D-1H (from running summary2csv after simulation, --time_index raw & daily):
![summary2csv_raw_WOPR](https://user-images.githubusercontent.com/46521401/96582925-42ef0d80-12dc-11eb-972c-cb214aec444b.JPG)
![summary2csv_daily_WOPR](https://user-images.githubusercontent.com/46521401/96583109-80ec3180-12dc-11eb-8849-f1997b099cf5.JPG)

It seems like the changes happens from after the date in question, not at the date. So is this just a plotting issue, or a one-day-shift-issue? What does ERT compare the observation at 14 Nov 1997 with? 4300 or 5586?

Same numbers for WOPRH:
![summary2csv_raw](https://user-images.githubusercontent.com/46521401/96583203-a6793b00-12dc-11eb-8102-034e544f754e.JPG)
![summary2csv_daily](https://user-images.githubusercontent.com/46521401/96583206-a8db9500-12dc-11eb-9a33-6ab5c3a7f018.JPG)

**Comment by anders-kiaer** [comment]
This is a well known challenge I think within AHM, which of course is most visible on large sudden changes in rates.

My experience:
* ERT will take what it gets from `UNSMRY` (i.e. what it gets when using `libecl`) on date specified in ERT obs.file.
* Schlumberger Eclipse (and probably also Flow?) will for rate vectors wait until next date when it comes to reporting used rate in `UNSMRY`. I.e. reported rate at time `t_i` in `UNSMRY` is within Eclipse used for `[t_(i-1), t_i)` (so for plotting purposes, it is usually most correct to _backfill_ from a date to previous date, for rate vectors). For cumulative vectors it is easier as it is more uniquely defined, there Eclipse simply reports the state of the cumulative vector at the given time.

Since it is obviously too late to change Eclipse behaviour :nerd_face: I guess the only "fix" would be that schedule files are created such that rate entered at date/time `t_i` is the one to be used for time interval `[t_i, t_i+1)` (as today?), but thanks to Eclipse, the created ERT observation file would need to shift its date one step into the future compared with schedule dates.

If I'm not too confused :confused: 

Maybe @eivindsm, @markusdregi or @asnyv has additional comments/experience/suggestions.

**Comment by eivindsm** [comment]
Yes, this is a well known issue.  @anders-kiaer is spot on, the rates specified in the WCONHIST section of eclipse and OPMFlow are target rates taking effect from the next day, the reporting dates in the schedule file however refers to the simulated production from that date particular, eg before the target rate takes effect. 

The E&P company that I work in has therefor issued a clear warning in their recommended practice document and advised against extracting observations from the schedule file. You will have to create a workaround, @wouterjdb

**Comment by asnyv** [comment]
From my experince I also think that @anders-kiaer is spot on here 👍 

But @eivindsm (might be that this is what you meant, and that you were only a bit unprecise in the chosen phrasing :wink:):
 
Just to add some details:
I am quite sure that the new target rate takes effect immediately after definition, not the next day (or report step)? (this is also how I interpret @anders-kiaer's time intervals above). It is only the reported values that are lagging behind, and reporting from the previous step. It is possible to define much finer steps in the Eclipse `DATES` keyword than a day, e.g. is this valid input as far as I know:

```
DATES
1 JAN 2000 00:00:00 /
/
```

So I believe the issue here is basically that you request a rate at a time where the rate is actually ill-defined, since it is at the instant the rate target changes, and in Eclipse it was apparently decided that they would return the previous value instead of the next. Maybe to make sure that they had a consistent way to return reasonable values at the last report step?

Taking a small next step would as far as I know report the new rate at a later time the same day. 

I also believe `libecl` makes this assumption for `EclSum`'s `numpy_vector`, even without including new report steps in the actual simulation.
E.g. if you set a new value JAN 1 2000 in Eclipse, requesting a rate value at `time_index=[datetime.datetime(2000,1,1,0,0)]` returns the previous value, whilst requesting a rate a second later at `time_index=[datetime.datetime(2000,1,1,0,1)]` returns the new value?
Note though that `libecl` only works on a list of mainly rate vectors (as far as I know), so it is risky to use/assume this behavior. The rest I believe are interpolated linearly between report steps, including e.g. WBHP which in my opinion should rather behave as an instant change like the rate.

**Comment by wouterjdb** [comment]
#58 was previous work on this issue.

**Comment by olwijn** [comment]
Result of a new check of this issue:
![image](https://user-images.githubusercontent.com/13539085/104034956-d7795580-51d1-11eb-9c7c-2e83dacbca9a.png)

**Comment by olwijn** [comment]
Schedule data will be extracted from the production data dataframe. The time entries are shifted back in time by one time step such that for example (measured or simulated) rates at the end of each time step appear as well targets for Flow at the start of that time step.

**Comment by olwijn** [comment]
![image](https://user-images.githubusercontent.com/13539085/105382143-0ea72800-5c10-11eb-84b4-f790e5ed00df.png)
I think the problem is now (temporarily) solved. The solution may not be ideal since it requires the creation of a second 'schedule' variable, one to create the FlowNet SCHEDULE section, and a second one to use in creating the observations. But it works.

**Comment by asnyv** [comment]
FYI: 
Quite sure this behavior doesn't apply to total/cumulative vectors, e.g. FOPT, FOIP and etc. They are handled by (at least Eclipse) as continuous vectors, so that the rate defined e.g. 2000-01-01 is actually used from 2000-01-01 to calculate the total over the next timestep. The point here is that Eclipse handles the new WCONHIST input as an instant shift, and if you define WCONHIST e.g. 2000-01-01; there are two states at the same time, the original one, and the updated. Eclipse has decided to report the original one, but for the continuous vectors like FOPT it doesn't matter, as the total doesn't change over the instant shift of WCONHIST.

By just shifting the historical data you are therefore introducing new errors...

**Comment by olwijn** [comment]
Not sure if I fully understand the raised concern, but I did 2 simulation just to check if prescribing rates through WCONPROD or WCONHIST makes any difference.
![image](https://user-images.githubusercontent.com/13539085/107652753-345ca580-6c81-11eb-82b2-ce68affe0c83.png)
![image](https://user-images.githubusercontent.com/13539085/107652792-3de60d80-6c81-11eb-8434-f9524009ec69.png)

**Comment by asnyv** [comment]
This I don't believe @olwijn 😅 (unless you have also shifted your inputdata with one step). 
If so, you should in any case not diverge from the standard Eclipse output, that will just create other issues and confusions.

The point is that this behavior is not an Eclipse bug or anything like that, it is just that Eclipse (and hopefully FLOW) reports the (constant) rates for the timestep t0->t1 as (t0, t1] and not [t0, t1), so it is the value at t1 (the end) which is the valid value for the timestep, and not the first one at t0. While you are probably using your input to WCONHIST/WCONINJH at t0 as an observation, and then you get a mismatch?

Tried to make a simple example of what I mean. The yellow dots are what you give at a timestep in the schedule (after the `DATES` keyword for that date). The green dots are the rates as reported in e.g. UNSMRY. The Orange line is how Eclipse behaves over the timesteps, and the blue is the cumulative (on the secondary axis).

My point is thus that I think you are trying to match observations as the yellow dots, with the reported values which are the green dots? Shifting the whole schedule with one timestep should would lift the orange line by 1000, and also affect your cumulative?

![image](https://user-images.githubusercontent.com/47146384/107657908-e72f0280-6c85-11eb-9c20-d857a5d6d1c5.png)

**Comment by olwijn** [comment]
Thanks @Atgeir for the example. The data that we extract from an ECLIPSE run for matching with FlowNet are the green dots in your figure, i.e. values reported by ECLIPSE at the initial time (0) and at the end of each simulation time step. In order to be able to match these data with FlowNet we have to make sure that these values are set as well targets for the Flownet simulation at the beginning of the time step (the yellow dots). So what we do when we create the FlowNet schedule is ignore the very first green dot at time 0 and shift the subsequent remaining green dots back in time by one time step, resulting in the yellow dots that are in your figure, which are used as the well targets that are used in the schedule. So based on the data from this example, at the initial time (01.01.2000) we would write a WCONPROD keyword with a target rate value of 1000. If Flownet is able to realize that target rate over the entire subsequent 10-day simulation interval, it would produce the same total/cumulative rate as reported by ECLIPSE at day 10.  Do you agree?

**Comment by asnyv** [comment]
Hmm, since I don't know any of the FlowNet stack, I am a bit confused: 
you basically say that you create the FlowNet _input_ the Eclipses imulation _output_? In that case, what you do makes sense (I think) as long as you only handle vectors like rates and network pressures (which is what you need for WCONXXXX), and not totals/cumulatives.

If the case is that you create a schedule based on Eclipse output, I have a "stupid question": Why not use the Eclipse input directly? Do you run the Eclipse simulation and then try to match that in FlowNet as the "truth"?

**Comment by olwijn** [comment]
Yes, we generate the FLowNet input from the ECLIPSE output instaed of from the ECLIPSE deck (schedule text file). This is because we ultimatley want to generate FlowNet directly from measured data in cases where we do not have an ECLIPSE schedule file.  We do not (at this moment) use totals/cumulatives to generate the FlowNet schedule, but we do have an option to match Flownet totals to ECLIPSE totals (but in the history matching workflow). So yes, your final sentence is correct. Hope this clarifies things,

**Comment by asnyv** [comment]
yes it does, sorry to disturb 😅
